### DISCUSSION TITLE
Use config from non-default branch (No Forking)

### DISCUSSION THREAD

**Discussion by AlexanderFuergut** [renovatebot/renovate#22264] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Most recent

### If you're self-hosting Renovate, select which platform you are using.

Azure DevOps (dev.azure.com)

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Have config changes (both config.js and renovate.json) on a branch (eg _/new_renovate_feature_) and run Renovate in our pipeline, so that it uses the config on the branch for that run instead of the config on the default branch.

### What you tried so far.

Using _baseBranches_ to set the baseBranch to the one with the config changes, but it doesn't use the config on the branch.
Using _force_ in the config.js, but since config.js is also not being used from the branch it doesn't do anything.

**Comment by rarkins** [comment]
If you configure your `config.js` to use `baseBranches=new_renovate_feature` and `useBaseBranchConfig=true` then that might achieve what you're after

**Reply by AlexanderFuergut** [reply]
Thanks. The content of the config.js on the branch is ignored in favor of the base branch, so doing changes there has no effect.
What worked was setting the baseBranch as an environment var and setting the use base branch config as a runtime parameter with --use-base-branch-config=merge

**Reply by rarkins** [reply]
There is no such thing as `config.js` in a branch. `config.js` is server-side ("global") config, while in the repo you have "repo" config such as `renovate.json`. It sounds like you're putting the wrong config file in the repo

**Reply by AlexanderFuergut** [reply]
What do you mean there is no config.js in a branch and it is "server side"? I followed the Azure DevOps getting started article where a config.js is created in the top level of the repository together with pipeline.yaml, .npmrc and renovate.json files: https://docs.renovatebot.com/modules/platform/azure/

**Reply by rarkins** [reply]
If you choose to have `config.js` on a branch, you need to check out that branch before you run Renovate. Renovate won't notice or care which branch it's running in, or even whether or not it's running in a git repo.

Typically you don't run Renovate 1:1 in a repo (one config.js and one renovate.json per repo). Instead you should have a single repo with config.js which then renovates all other repos you want

**Comment by exoszajzbuk** [comment]
The issue is valid with `renovate.json` as well, or maybe it's just not clear for me how the "useBaseBranchConfig: merge` should work.

I have the following (basically empty) `renovate.json` on the `main` branch:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json"
}
```

I "extend" `renovate.json` on `test-branch` with the following:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:base"]
}
```

so if I run `renovate` with the following env vars (a.k.a these would be in the central `config.js`):

```yaml
    RENOVATE_USE_BASE_BRANCH_CONFIG: 'merge'
    RENOVATE_BASE_BRANCHES: '["test-branch"]'
```

then I can't see `config:base` getting applied, e.g. no "Dependecy Dashboard" issue is getting create. Or if I set `renovate.json` in the side branch `test-branch` to

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [":disableRenovate"]
}
```

I'd expect Renovate to be completely disabled "on that branch", but it isn't happening, it seems like the main branch config is taken into consideration in this case as well

**Reply by exoszajzbuk** [reply]
I think the authors of https://github.com/renovatebot/renovate/discussions/19916 and https://github.com/renovatebot/renovate/discussions/16108 had the same intention in their minds: to have the ability to "test" changes in `renovate.json` on side branches (most probably combined with a `--dry-run` flag to make sure no "test" issues and PRs are getting created)

For this kind of workflow the `useBaseBranchConfig: merge` strategy is probably _inherently flawed_, as it has several problems:

1. With no `renovate.json` present on the `main` branch, the whole merging process fails right away, but it should not be the case.
2. The "merging" logic would make it really hard to test scenarios of **removing** or **replacing** rules.

Handling the merging of config files across branches is a nice feature, but probably not totally necessary in an inherently Git-based environment. I should be able to for example:

1. add/remove/update rules in the repo's `renovate.json` freely in side-branches and then do a dry-run to see the results before merging the changes and potentially affect the "real" scheduled update jobs running on the `main` branch
2. just move the config file from one valid location to another, like from `renovate.json` to `.gitlab/renovate.json` and see with a dry-run on the side branch that the dependencies are not affected at all
3. like the previous point, to change the format of `renovate.json` to e.g. `renovate.json5` to be able to add comments, etc and again, to verify with a dry-run that this cosmetic change done right does not affect anything

---

I obviously can't yet see deep into the details of the implementation, but what goes agains having a `useBaseBranchConfig: override` setting, which just completely ignores whatever config file is (or isn't) on the `main` branch, seaches for a valid config file on each `baseBranches` and leaves handling the merges to the maintainer of the repository?

Theoretically it should not affect any priorities between server config, env vars, presets, command line arguments, it's just about taking the entirety of `renovate.json` (or any valid location) from the `baseBranch` itself.

**Reply by reduckted** [reply]
> I obviously can't yet see deep into the details of the implementation, but what goes agains having a `useBaseBranchConfig: override` setting, which just completely ignores whatever config file is (or isn't) on the main branch, seaches for a valid config file on each baseBranches and leaves handling the merges to the maintainer of the repository?

@rarkins Would  you be open to adding an `override` option for `useBaseBranchConfig` (or something similar)? As @exoszajzbuk points out, this kind of question has been asked a few times, so there's definitely a use case for it. I'd be willing to help contribute if necessary.

**Reply by exoszajzbuk** [reply]
@reduckted for adding the `replace` option (with the proposed "override" functionality) a PR was already drafted by me: https://github.com/renovatebot/renovate/pull/22658 but got closed due to inactivity

I did not have time to figure out this bit: https://github.com/renovatebot/renovate/pull/22658/files#diff-34f2d644001d505d48bb39f08261487e687cabdc839943de43f4369ab974f024R86 about how to detect a (possibly modified) config file name on the given "base branch" - enforcing the same filename would defeat half of the purpose of this `replace` option

if you can help by picking this up, the only other thing missing is to align the proposed test case in the PR

**Reply by reduckted** [reply]
I'll take a look and see what I can do, thanks. 👍

**Reply by reduckted** [reply]
It ended being more complicated than just _replacing_ the default config, so I started fresh.

https://github.com/renovatebot/renovate/pull/36084

**Reply by reduckted** [reply]
_Replacing_ the config from the default branch is basically impossible (see https://github.com/renovatebot/renovate/pull/36084#issuecomment-2904291820), and I've encountered a scenario where, even if it were possible to use a config from another branch, it doesn't necessarily help you test a config from that other branch.

The scenario that I had was I needed to change some package rules, but also add a `postUpdateCommand` to some of them that would run a script. I wanted to test those changes before they were merged into the default branch. The problem is, that script that I wanted to run _was not_ on the default branch. So even if renovate could use a _config_ from another branch, it wouldn't help because the _script_ that the config wants to run isn't available.

The only way that scenario could be solved is by actually running Renovate on that new branch.

@rarkins, @viceice Would you consider an option to specify the branch that Renovate runs on? I've made an attempt at doing that, and it seems to work for GitHub. I can make a draft PR if you'd like to take a look.

**Reply by rarkins** [reply]
Specifying "the branch that Renovate runs on" means it would need to be specified as part of the global config? And most likely this means you run on only one repo?

**Reply by reduckted** [reply]
That's correct. For my use-case I use the npm package and run it on a single repository per execution.

**Reply by reduckted** [reply]
🤔 As a possible alternative, since I only want to _test_ configuration changes, would using `platform=local` and checking out the repository on the specific branch manually be a suitable alternative? Aside from what's listed in [the docs](https://docs.renovatebot.com/modules/platform/local/), is there anything else that I might need to watch out for?

**Reply by rarkins** [reply]
When you say _test_ configuration changes, you mean beyond just validation and instead knowing which PRs would be opened or closed, etc?

**Reply by reduckted** [reply]
Yes, essentially I want to confirm that I've defined the package rules correctly such that:
- dependencies that shouldn't be updated won't be.
- dependencies that should be updated will be.
- pull requests would be grouped how I want them to be.
- files modified by post-upgrade commands are found correctly (i.e. are the `fileFilters` defined correctly)

And doing all of that without actually creating any pull requests - i.e perform a dry run.

**Reply by rarkins** [reply]
@RahulGautamSingh how close did you get to this in your "Reconfigure Renovate" PR work? e.g. being able to "simulate" the new run and describe changes?

**Reply by RahulGautamSingh** [reply]
Very close, to what you explain ie. simulate run & describ changes since we do extraction + lookup (branchifyUpgrades). 

As pe above comment, I think:

1. dependencies that shouldn't be updated won't be. ✅
2. dependencies that should be updated will be. ✅
3. pull requests would be grouped how I want them to be -> (not tested)
4. files modified by post-upgrade commands are found correctly (i.e. are the fileFilters defined correctly) -> (not tested)

I will test the remaining two cases and update here.

**Reply by reduckted** [reply]
> As a possible alternative, since I only want to test configuration changes, would using `platform=local` and checking out the repository on the specific branch manually be a suitable alternative?

So I've given this a try and it actually works quite well. I only have two complaints:

1. It logs _warnings_ about not being able to sync branches because it's using the "local" platform. It would be nice if there was some way to suppress those warnings.
2. I have to use debug logging to find out what would be changed, but it sounds like the work @RahulGautamSingh is doing might solve that.

All in all, I'm pretty happy to use `platform=local` for testing/verifying config changes, because it's not something that I have to do very often.

@RahulGautamSingh Are the changes you're working on available somewhere? I'm interested in seeing what they are.

**Reply by RahulGautamSingh** [reply]
Here you go: https://github.com/renovatebot/renovate/pull/34696

**Reply by RahulGautamSingh** [reply]
Tested case 3: https://github.com/Rahul-renovate-testing/interactive-grouped/pull/3/files

Works well

Case 4, doesn't seem to be a possible case for the reconfigure pr logic as we do not go beyond branchifying the upgrades. wdyt? @rarkins

**Reply by reduckted** [reply]
> It logs _warnings_ about not being able to sync branches because it's using the "local" platform. It would be nice if there was some way to suppress those warnings.

I looked into this further, and it's actually not really a warning - it's an error that causes Renovate to terminate: https://github.com/renovatebot/renovate/discussions/36538

**Reply by rarkins** [reply]
At this point we don't plan to run a _full_ dry run for the reconfigure branch. It would approximately double the runtime on a repo plus introduce a lot of challenging edge cases (e.g. there might be differences in files because the reconfigure branch is not up to date with the default branch)
